metrics.go

```diff
@@ -76,3 +121,63 @@ func must[V any](v V, err error) V {
 	}
 	return v
 }
+
+// sampleSet stores a bounded set of samples and exposes the ability to check
+// whether it contains a given sample. See sampleSet.contains.
+//
+// The memory footprint of sampleSet is around 2 * maxSize * 96 bytes.
+type sampleSet struct {
+
+	// We can use existing LRU implementations for this at the price of slightly
+	// higher memory footprint and explanation that recency is unused. Hence the
+	// hand-rolled data structure here.
+
+	// maxSize defines the maximum number of samples to store.
+	maxSize int
+	// mu protects access to samples.
+	mu sync.Mutex
+	// samples stores the samples using a map for quick lookup.
+	samples map[string]struct{}
+	// order maintains the order of samples to evict by oldest entry irrespective of
+	// recency.
+	order *list.List
```

Comment:
This works but, for this kind of thing, I generally prefer a live set and a decaying set. I.e.:

1. Fill up a live set until we reach a threshold.
2. At that point, rotate it out into the "decaying slot", never update it, discarding the previous decaying set.
3. Create a new live set.

It simplifies GC and removes the need for an ordered list, at the cost of potentially 2x memory.

Reply:
I like this; this approach should keep the memory footprint at 2X ish while potentially doubling the sample size 🍻

Reply:
@Stebalien I have implemented this suggestion. would you mind giving this PR another look before I merge please?

Reply:
Actually hold off. I have missed something.

Reply:
> I have missed something.

For reference, this was storing half the max size in each subset map, which is not right since it cannot guarantee uniqueness check across _at least_ max size number of samples. I changed it to use the max size directly as the size of each subset.